Use latest version of virtualenv

benjyw opened the pull request:
For compatibility with Python3.9.  We now consume virtualenv as a zipapp, as the previous method no longer works.

We also now check virtualenv's sha256, for added security. 
This change also fixes some cases where we weren't exiting an outer process when a subprocess exited on error.

benjyw committed:
Simplify creating of the cached Pants venv.

Previously we bootstrapped the virtualenv tool from PyPI. But:
A) This adds complexity.
B) There are issues with virtualenvs thus created on
   brew-installed Python interpreters on MacOS.

This change switches to using the built-in `venv` module, which
has a subset of full virtualenv functionality that is sufficient
for our needs.

benjyw commented:
@jsirois checking that this is OK, in case we bootstrapped the full `virtualenv` tool for a reason.

Eric-Arellano reviewed (approved):
I tried doing this in https://github.com/pantsbuild/setup/pull/70 and I gave up because `venv` wasn't pre-installed in some environments. But imo, this is a good change to make and we should have landed #70 last year.

This has an extra benefit: one fewer download to make, improving our supply chain security.

Eric-Arellano commented:
> in case we bootstrapped the full virtualenv tool for a reason.

Mostly legacy, other than issue in #70. Python 2 did not have builtin venv.

jsirois commented:
Yeah, this will need to be more sophisticated. Beyond Python 2 needing virtualenv we have the fairly common problem of Linux distributions that do not include the venv module by default. Debian distros do this and this is similar to https://github.com/pantsbuild/pex/issues/1027. I think we either need to use a more modern virtualenv or else try `-mvenv` and either fall back to virtualenv or else, similar to Pex, link an issue that explains why the venv module might be missing / how to get it in some known cases.

Eric-Arellano commented:
> Beyond Python 2 needing virtualenv

To be clear, that's a non-issue for Pants. This script has long assumed you're running Pants with Py3.

jsirois commented:
Ack. But also to be clear, the missing venv module under Python 3 is real. As such we were correct to never land #70.

benjyw commented:
Hrmph, OK, will think on this.

benjyw commented:
So, upgrading virtualenv leads to other complications: the latest versions no longer have the root `virtualenv.py` script. They now strongly expect you to install via pip or similar. Which leads to a whole other can of worms, since we don't know if we're even able to do so. 

So given this, plus the fact that we want the setup script to be as simple as possible, one option is to require venv to be available in the python interpreter, check for this, and give a useful error message. We already require an interpreter, so requiring one with minimal functionality doesn't seem awful. Thoughts?

Longer term, packaging Pants with its interpreter (a la PyOxidizer or similar) seems like the way to go.

jsirois commented:
Works for me I guess. As long as we acknowledge we're solving one user's issue and opening up a new missing venv issue for pre-existing users on their next Pants version bump in CI (I assume the majority of Debian use cases are CI).

benjyw commented:
> Works for me I guess. As long as we acknowledge we're solving one user's issue and opening up a new missing venv issue for pre-existing users on their next Pants version bump in CI (I assume the majority of Debian use cases are CI).

Well, only if they upgrade the script, right? Just upgrading Pants with the existing runner script should be OK (until they try to upgrade to Python 3.9).

At present we're seeing at least two new users hit issues with this current script and Python 3.9, and using a newer virtualenv seems tricky, unless we're able to pip-install it. Which we could make a best-effort attempt to do I guess?

Eric-Arellano commented:
> Well, only if they upgrade the script, right?

Right

jsirois commented:
Correct, but they need to upgrade the script if they have users bitten by the bug you're fixing here. There is a push-pull. There is 1 pants script per repo. That repo will be used by desktop users and CI. It would be awesome not to force a repo to pick one of these to not break and I stead not break either.

benjyw commented:
Fair point. So should we try to directly pip-install virtualenv? What if `pip` isn't present? It does seem like recent versions of virtualenv cannot be downloaded and run directly, as we do today.

jsirois commented:
Well, simply using the status quo virtualenv as a fallback would strictly move us forward. We'd cover all the old cases as well as the motivating issue here.

That said, did you check out the virtualenv installation docs? https://virtualenv.pypa.io/en/latest/installation.html#via-zipapp

benjyw commented:
I did not read them in enough detail! I didn't know what zipapp was but assumed it was some tool we would need to bootstrap in turn. This is promising.

benjyw committed:
Use new virtualenv via zipapp.

But use venv if possible.

benjyw commented:
OK, made a change that I think combines the best of all worlds. LMKWYT.

jsirois commented on pants:
```diff
@@ -34,10 +34,12 @@ fi
+# the lowest version of Python we support running Pants on.
+# See https://virtualenv.pypa.io/en/latest/installation.html#via-zipapp
+VIRTUALENV_ZIPAPP=virtualenv.pyz
+VIRTUALENV_ZIPAPP_URL="https://bootstrap.pypa.io/virtualenv/3.6/${VIRTUALENV_ZIPAPP}"
```
The lowest version we currently support, IIUC, is 3.7. Is 3.6 here to support older versions of Pants? If so can you note how old?

benjyw replied:
Yep, see https://github.com/pantsbuild/setup/blob/gh-pages/pants#L125  which has logic to map pants versions to supported python versions. 

But note that the current virtualenv release still supports all the way back to 2.7, so we're good for a while...

jsirois commented on pants:
```diff
@@ -250,12 +253,25 @@ function bootstrap_pants {
       local staging_dir
       staging_dir=$(tempdir "${PANTS_BOOTSTRAP}")
+
+      if ${python} -c "import venv" > /dev/null  2>&1; then
```
I think you want to quote `${python}` - people do strange things.

benjyw replied:
Done.

jsirois commented on pants:
```diff
@@ -250,12 +253,25 @@ function bootstrap_pants {
+        virtualenv_cmd="${python} -m venv"
+      else
+        # The venv module is not available, so download and use the virtualenv zipapp.
+        green "${python} does not have the venv module avilable, using virtualenv instead."
```
Spelling on available.

benjyw replied:
Fixed.

benjyw committed:
Add SHA256 checking for virtualenv.

Also ensure that "die" does in fact abort the script
when called in a subshell.

benjyw committed:
Address code review

benjyw commented:
Addressed the comments, and also added SHA256 checking for the virtualenv zipapp. Might as well.

jsirois commented on pants:
```diff
@@ -197,20 +201,25 @@ function determine_python_exe {
-      mv "${staging_dir}/latest" "${PANTS_BOOTSTRAP}/${VENV_PACKAGE}"
-    ) 1>&2
+      curl -LO "${VIRTUALENV_ZIPAPP_URL}"
+      fingerprint="$(shasum -a 256 "${VIRTUALENV_ZIPAPP}" | cut -d' ' -f1)"
```
Hrm. Do we have this dependency already? We could just use python for this since we already depend on it.

benjyw replied:
Yeah good point. Might as well.

benjyw replied:
Done.

benjyw committed:
Use python to compute sha256.

So we don't have to assume existence of shasum and cut.

benjyw committed:
Lint

Eric-Arellano commented on pants:
```diff
@@ -34,10 +34,13 @@ fi
-
-VENV_PACKAGE=virtualenv-${VENV_VERSION}
-VENV_TARBALL=${VENV_PACKAGE}.tar.gz
+# Note that this URL points to the zipapp for the latest version of virtualenv
```
> for the latest version of virtualenv

Is the SHA going to be stable then? If a new release comes out, I'd expect that to change and break people's builds.

benjyw replied:
Oh you make an excellent point. It will break on the next virtualenv upgrade. Damn. Good catch!

benjyw replied:
Well, crap. There is no way that I can see to download a zipapp at a specific version. It's like they actively don't want you to verify these.

benjyw replied:
Oh, there is: ```If you are looking for past version of virtualenv.pyz they are available here: https://github.com/pypa/get-virtualenv/blob/<virtualenv version>/public/<python version>/virtualenv.pyz?raw=true```

benjyw replied:
Phew

Eric-Arellano replied:
> If you are looking for past version of virtualenv.pyz they are available here: https://github.com/pypa/get-virtualenv/blob/<virtualenv version>/public/<python version>/virtualenv.pyz?raw=true

Eric-Arellano commented on pants:
```diff
@@ -194,23 +198,44 @@ function determine_python_exe {
+      curl -LO "${VIRTUALENV_ZIPAPP_URL}"
+      fingerprint="$(compute_sha256 "${python}" "${VIRTUALENV_ZIPAPP}")"
+      if [[ "${VIRTUALENV_ZIPAPP_EXPECTED_SHA256}" != "${fingerprint}" ]]; then
+        die "SHA256 of ${staging_dir}/${VIRTUALENV_ZIPAPP} is not as expected. Aborting."
```
Maybe explicitly reference `VIRTUALENV_ZIPAPP_URL`? As a user, I'm not sure I would understand how to fix this issue.

Eric-Arellano reviewed (commented):
Thanks!

benjyw committed:
Fix the virtualenv zipapp version

benjyw committed:
Don't try to use the venv module after all.

In some cases, on Ubuntu it fails to install pip into the virtual
environment. virtualenv is more robust, and this is slightly simpler
too.

benjyw commented:
OK a few more changes:

- Get a fixed version of virtualenv.
- Don't even bother trying to use venv if available, as it turns out (see Travis failures) that it doesn't always install pip into the virtual environment, even though it is supposed to. virtualenv doesn't appear to have this problem (as I tested by sshing into the Travis container and trying things out). 
- A few improvements of var assignments and stuff.

Eric-Arellano reviewed (approved):
Excellent, thanks Benjy

jsirois commented on pants:
```diff
@@ -245,32 +267,35 @@ function bootstrap_pants {
+      green "Installing ${pants_requirement} into a virtual environment at ${bootstrapped}"
       # shellcheck disable=SC2086
-      "${python}" "${venv_path}/virtualenv.py" --no-download "${staging_dir}/install" && \
+      ${python} "${virtualenv_path}" --no-download "${staging_dir}/install" && \
```
The quoting of `${python}` was correct. `/Users/Jane Doe/...` as a home for python is both valid and likelyish.

benjyw committed:
Quote python

benjyw commented:
@jsirois do these errors ring any bells? https://travis-ci.com/github/pantsbuild/setup/jobs/508731786

jsirois commented:
Nope. But I fiddled. Its not pyenv shims which was my 1st suspect. It is the symlink to the zippapp. I can repro only when I set up one of those and use it:
```
~/Downloads $ mkdir -p some/other/dir
~/Downloads $ ln -s $PWD/virtualenv.pyz some/other/dir/
~/Downloads $ ls -l some/other/dir/
total 0
lrwxrwxrwx 1 jsirois jsirois 38 May 27 13:43 virtualenv.pyz -> /home/jsirois/Downloads/virtualenv.pyz
~/Downloads $ python some/other/dir/virtualenv.pyz bob
Traceback (most recent call last):
  File "/home/jsirois/.pyenv/versions/3.8.1/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/jsirois/.pyenv/versions/3.8.1/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "some/other/dir/virtualenv.pyz/__main__.py", line 168, in <module>
  File "some/other/dir/virtualenv.pyz/__main__.py", line 164, in run
  File "some/other/dir/virtualenv.pyz/virtualenv/__main__.py", line 18, in run
  File "some/other/dir/virtualenv.pyz/virtualenv/run/__init__.py", line 32, in cli_run
  File "some/other/dir/virtualenv.pyz/virtualenv/run/session.py", line 46, in run
  File "some/other/dir/virtualenv.pyz/virtualenv/run/session.py", line 53, in _create
  File "some/other/dir/virtualenv.pyz/virtualenv/create/creator.py", line 171, in run
  File "some/other/dir/virtualenv.pyz/virtualenv/create/via_global_ref/builtin/via_global_self_do.py", line 101, in create
  File "some/other/dir/virtualenv.pyz/virtualenv/create/via_global_ref/api.py", line 89, in create
  File "some/other/dir/virtualenv.pyz/virtualenv/create/via_global_ref/api.py", line 92, in install_patch
  File "some/other/dir/virtualenv.pyz/virtualenv/create/via_global_ref/builtin/cpython/cpython3.py", line 27, in env_patch_text
  File "some/other/dir/virtualenv.pyz/virtualenv/create/via_global_ref/api.py", line 103, in env_patch_text
  File "/home/jsirois/.pyenv/versions/3.8.1/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "some/other/dir/virtualenv.pyz/virtualenv/app_data/base.py", line 58, in ensure_extracted
  File "/home/jsirois/.pyenv/versions/3.8.1/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "some/other/dir/virtualenv.pyz/virtualenv/app_data/via_disk_folder.py", line 83, in extract
  File "some/other/dir/virtualenv.pyz/virtualenv/util/zipapp.py", line 22, in extract
    
  File "/home/jsirois/.pyenv/versions/3.8.1/lib/python3.8/zipfile.py", line 1439, in getinfo
    raise KeyError(
KeyError: "There is no item named 'virtualenv.pyz/virtualenv/create/via_global_ref/_virtualenv.py' in the archive"
```

benjyw committed:
Don't symlink the virtualenv zipapp.

It causes it to fail in some cases. And we can simply
mv the file, which is just as atomic.

benjyw commented:
OK, one more change to get rid of the symlink. It seems unnecessary anyway, and I think I just cargo-culted it.

jsirois commented on pants:
```diff
@@ -228,8 +228,8 @@ function bootstrap_virtualenv {
       green "SHA256 fingerprint of ${VIRTUALENV_ZIPAPP_URL} verified."
-      ln -s "${staging_dir}/${VIRTUALENV_ZIPAPP}" "${staging_dir}/latest"
-      mv "${staging_dir}/latest" "${bootstrapped}"
+      mv -f "${staging_dir}/${VIRTUALENV_ZIPAPP}" "${bootstrapped}"
```
I just noticed this but `VIRTUALENV_ZIPAPP` is just the unqualified `virtualenv.pyz`; so bumping the version we download / changing the expected sha256 going forward will no-op above on line 218. So that needs a fix. At that point it may make sense to house the downloaded `virtualenv.pyz` in a directory with its version or hash or both as part of the naming scheme. Then here make a hard link instead of doing a mv.
This way multiple clones of different pants-using repos won't thrash the cache.

benjyw replied:
Oh, good catch. Not sure why we need the hard link, if the final resting place of the file, as returned from this function, includes the version? This mv and rmdir is just to clean up the tmpdir. See most recent commit for a fix for this issue, does that seem reasonable?

benjyw committed:
Ensure virtualenv versions don't conflict.

Also fix a file check that was mistakenly a directory check.

jsirois reviewed (approved):
Yeah, I think this probably works.

I'm never quite sure on the atomicity of `mv` on various OSs / distros. I'm pretty positive an `ln -f` followed by a remove of the original structure is always atomic though.

benjyw commented:
Thanks for the eagle-eyed reviews and great feedback! Phew...
